notebooks/bellingcat/search-grid-generator.ipynb

```diff
@@ -0,0 +1,136 @@
+    "\n",
+    "# @markdown Customise the name of the output file:\n",
+    "file_name = \"grid.kml\"  # @param {type:\"string\"}\n",
+    "\n",
+    "import numpy as np\n",
+    "from ipywidgets import widgets\n",
+    "from IPython import get_ipython\n",
+    "from IPython.display import display\n",
+    "from datetime import datetime\n",
+    "\n",
+    "\n",
+    "logfile = f\"./grid_log_{datetime.now().strftime('%d%m%Y_%H%M%S')}.log\"\n",
+    "\n",
+    "# Install dependencies and enable widgets if running in colab\n",
+    "if \"google.colab\" in str(get_ipython()):\n",
+    "    !jupyter nbextension enable --py widgetsnbextension >> {logfile} 2>&1\n",
+    "    ![ ! -f \"deps_installed\" ] && pip install simplekml >> {logfile} 2>&1 && touch \"deps_installed\"\n",
+    "\n",
+    "import simplekml\n",
+    "\n",
+    "\n",
+    "def generate_grid(\n",
+    "    lat1, lon1, lat2, lon2, lat_step=0.01, lon_step=0.01, file_name=\"grid.kml\"\n",
+    "):\n",
+    "    # Create a kml object\n",
+    "    kml = simplekml.Kml()\n",
+    "    # create a folder\n",
+    "    fol = kml.newfolder(name=\"Grid\")\n",
+    "\n",
+    "    for lat in np.arange(lat1, lat2, lat_step):\n",
+    "        for lon in np.arange(lon1, lon2, lon_step):\n",
+    "            # Round the lat and lon values to 5 decimal places\n",
+    "\n",
+    "            # Create a polygon\n",
+    "            pol = fol.newpolygon(name=f\"({lat}, {lon})\")\n",
+    "            pol.outerboundaryis = [\n",
+    "                (lon, lat),\n",
+    "                (lon + lon_step, lat),\n",
+    "                (lon + lon_step, lat + lat_step),\n",
+    "                (lon, lat + lat_step),\n",
+    "                (lon, lat),\n",
+    "            ]\n",
+    "            pol.style.linestyle.color = simplekml.Color.white\n",
+    "            pol.style.linestyle.width = 2\n",
+    "            # make the polygon transparent\n",
+    "            pol.style.polystyle.color = simplekml.Color.changealphaint(\n",
+    "                0, simplekml.Color.red\n",
+    "            )\n",
+    "\n",
+    "    # Save the kml file\n",
+    "    kml.save(file_name)\n",
+    "\n",
+    "\n",
+    "# Add kml extension to the file name if it doesn't exist\n",
+    "if not file_name.endswith(\".kml\"):\n",
+    "    file_name += \".kml\"\n",
+    "\n",
+    "# Extract the latitude and longitude\n",
+    "lat1, lon1 = lat_long_1.replace(\" \", \"\").split(\",\")\n",
+    "lat2, lon2 = lat_long_2.replace(\" \", \"\").split(\",\")\n",
+    "\n",
+    "# Reorder the lat long values\n",
+    "lat1, lat2 = min(float(lat1), float(lat2)), max(float(lat1), float(lat2))\n",
+    "lon1, lon2 = min(float(lon1), float(lon2)), max(float(lon1), float(lon2))\n",
+    "\n",
+    "# Convert grid_size to degrees for lat_step and lon_step separately\n",
+    "lon_step = grid_size / 111\n",
+    "lat_step = grid_size / 111 * np.cos(np.radians((lat1 + lat2) / 2))\n",
```

Comment:
This is calculated incorrectly, leading to inconsistent grid sizes. **The length of a degree of latitude is invariant, not longitude** -- lines of longitude get closer at the poles. Should be something like: 

```
lat_step = grid_size / 111
lon_step = grid_size / (111 * np.cos(np.radians((lat1 + lat2) / 2)))
```

Reply:
Ah good spot, thanks! I had checked that the grid was square, but not that the cells were correctly sized!

fixed in [4215965](https://github.com/bellingcat/open-source-research-notebooks/pull/20/commits/42159657eb00d1adc4fc75784fd5ab05ccefedf2)